pytorch_lightning/loops/training_epoch_loop.py

```diff
@@ -0,0 +1,422 @@
+
+            if training_epoch_end_output is not None:
+                raise MisconfigurationException(
+                    'training_epoch_end expects a return of None. '
+                    'HINT: remove the return statement in training_epoch_end'
+                )
+
+        # call train epoch end hooks
+        self._on_train_epoch_end_hook(processed_outputs)
+        self.trainer.call_hook('on_epoch_end')
+        self.trainer.logger_connector.on_epoch_end()
+        return self.epoch_output
+
+    def _on_train_epoch_end_hook(self, processed_epoch_output: List[List[STEP_OUTPUT]]) -> None:
+        """Runs ``on_train_epoch_end hook``."""
+        # We cannot rely on Trainer.call_hook because the signatures might be different across
+        # lightning module and callback
+        # As a result, we need to inspect if the module accepts `outputs` in `on_train_epoch_end`
+
+        # This implementation is copied from Trainer.call_hook
+        hook_name = "on_train_epoch_end"
+        prev_fx_name = self.trainer.lightning_module._current_fx_name
+        self.trainer.lightning_module._current_fx_name = hook_name
+
+        # always profile hooks
+        with self.trainer.profiler.profile(hook_name):
+
+            # first call trainer hook
+            if hasattr(self.trainer, hook_name):
+                trainer_hook = getattr(self.trainer, hook_name)
+                trainer_hook(processed_epoch_output)
+
+            # next call hook in lightningModule
+            model_ref = self.trainer.lightning_module
+            if is_overridden(hook_name, model_ref):
+                hook_fx = getattr(model_ref, hook_name)
+                if is_param_in_hook_signature(hook_fx, "outputs"):
+                    self.warning_cache.warn(
+                        "The signature of `ModelHooks.on_train_epoch_end` has changed in v1.3."
+                        " `outputs` parameter has been deprecated."
+                        " Support for the old signature will be removed in v1.5", DeprecationWarning
+                    )
+                    model_ref.on_train_epoch_end(processed_epoch_output)
+                else:
+                    model_ref.on_train_epoch_end()
+
+            # call the accelerator hook
+            if hasattr(self.trainer.accelerator, hook_name):
+                accelerator_hook = getattr(self.trainer.accelerator, hook_name)
+                accelerator_hook()
+
+        # restore current_fx when nested context
+        self.trainer.lightning_module._current_fx_name = prev_fx_name
+
+    def _num_training_batches_reached(self, is_last_batch: bool = False) -> bool:
+        """Checks if we are in the last batch or if there are more batches to follow."""
+
+        # TODO: Can we combine this with training_batch_loop's arg that does a similar check?
+        return self.batches_seen == self.trainer.num_training_batches or is_last_batch
+
+    # TODO(@awaelchli): merge with on_advance_end()
```

Comment:
Seeing lots of TODOs scattered.

Was wondering whether they are things to fix in this PR (or a close next one)
Or things that could get done eventually but which are not that important (okay to get forgotten).

This is to me the distinction between a FIXME and a TODO

Reply:
@carmocca after introducing a general interface (this PR + follow up for evaluation/prediction) we want to revisit each of those loops individually to tackle those TODOs.

In this PR and the general interface follow-ups we only aim to introduce the interface to all loops and reused existing code whereever possible to limit the amount of code we need to care for and also make it easier to keep in sync.

Reply:
There are many todos (without awaelchli) that come from the old loop.

this particular one here, if you want, I can do it in this PR. 
def on_train_batch_end() comes from the old loop and while it can be integrated with on_advance_end() I still kept it here simply for sanity, in the case master changes this function I would have to search a long time where to add the changes. any hard ones that require substantial logic changes I prefer to not do in this PR.

Reply:
I can go over fixme vs. todo